Sources/TensorFlow/Core/LazyTensorTrace.swift

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CTensorFlow
+
+/// A class describing the tensor operations that need to be executed
+/// to evaluate a given `LazyTensorOperation`.
+class LazyTensorTrace {
+    // inputs will be "placeholder" nodes.
+    var inputs: [LazyTensorOperation] = []
+    var inputValues: [TFETensorHandle] = []
+    var operations: [LazyTensorOperation] = []
+    var outputs: [LazyTensorOperation] = []
+    var originalOutputs: [LazyTensorOperation] = []
+
+    init(_ lazyOp: LazyTensorOperation) {
+        // TODO: We only pick operations on which `lazyOp` depends on. Note that
+        // there may be other live tensors that could also be materialized at
+        // this time. e.g.,
+        //   x = a + b
+        //   y = x + c
+        // For `x`, only `a + b` is extracted. One optimization is to also include
+        // `y = x + c` into the trace so that we don't have the overhead of creating
+        // another trace when we need to materialize `y`.
+        //
+        _ = collectLazyOp(lazyOp)
+        lazyOpsCache.removeAll()
+    }
+
+    var signature:  String {
+        let inputsDesc = inputs.map { input -> String in
+            let dtypeAttr = input.attrs["dtype"]!
+            return "\(input.outputName): \(dtypeAttr)"
+        }
+        let inputDesc = inputsDesc.joined(separator: ", ")
+        let outputsDesc = outputs.map { "\($0.outputName)" }
+        let outputDesc = outputsDesc.joined(separator: ", ")
+        return "lazyTrace_\(operations.count)(\(inputDesc)) -> (\(outputDesc))"
+    }
+
+    private var lazyOpsCache: [ObjectIdentifier: LazyTensorOperation] = [:]
+
+    private func updateCacheAndOperations(
+        _ id: ObjectIdentifier, _ node: LazyTensorOperation) {
```

Comment:
```suggestion
        _ id: ObjectIdentifier, _ node: LazyTensorOperation
    ) {
```

Reply:
Consider compensating for weak type information for readability. It's better to label the first argument to describe _what_ object identifier it is.

Reply:
Renamed the function to make it a bit more readable.